File: src/geometry_area_of_polygons_medium.py
"""
geometry - area_of_polygons (medium)
Generated: 2026-02-22T04:12:17.262019
"""

from problem_utils import *
from svg_utils import Diagram

def generate():
    problem_type = choice(['rectangle', 'triangle', 'parallelogram', 'trapezoid', 'composite'])
    
    if problem_type == 'rectangle':
        length = randint(5, 20)
        width = randint(3, 15)
        area = length * width
        
        d = Diagram(width=300, height=200, padding=40)
        points = [(0, 0), (length, 0), (length, width), (0, width)]
        d.polygon(points)
        d.segment_label((0, 0), (length, 0), f"{length}")
        d.segment_label((length, 0), (length, width), f"{width}")
        d.right_angle((0, 0), (length, 0), (0, width))
        d.right_angle((length, 0), (length, width), (0, 0))
        
        return problem(
            question=f"Find the area of a rectangle with length ${length}$ units and width ${width}$ units.",
            answer=area,
            difficulty=(1200, 1300),
            topic="geometry/area_of_polygons",
            solution=steps(
                f"Area of rectangle = length × width",
                f"$A = {length} \\times {width}$",
                f"$A = {area}$ square units"
            ),
            image=d.render()
        )
    
    elif problem_type == 'triangle':
        base = randint(6, 24)
        height = randint(4, 20)
        area = Rational(base * height, 2)
        
        d = Diagram(width=300, height=220, padding=40)
        points = [(0, 0), (base, 0), (base/2, height)]
        d.polygon(points)
        d.segment_label((0, 0), (base, 0), f"{base}")
        d.line((base/2, 0), (base/2, height), dashed=True)
        d.segment_label((base/2, 0), (base/2, height), f"{height}")
        d.right_angle((base/2, 0), (base, 0), (base/2, height))
        
        return problem(
            question=f"Find the area of a triangle with base ${base}$ units and height ${height}$ units.",
            answer=area,
            difficulty=(1300, 1400),
            topic="geometry/area_of_polygons",
            solution=steps(
                f"Area of triangle = $\\frac{{1}}{{2}} \\times$ base $\\times$ height",
                f"$A = \\frac{{1}}{{2}} \\times {base} \\times {height}$",
                f"$A = \\frac{{{base * height}}}{{2}}$",
                f"$A = {latex(area)}$ square units"
            ),
            image=d.render()
        )
    
    elif problem_type == 'parallelogram':
        base = randint(8, 20)
        height = randint(5, 15)
        slant = randint(2, 6)
        area = base * height
        
        d = Diagram(width=320, height=200, padding=40)
        points = [(slant, 0), (base + slant, 0), (base, height), (0, height)]
        d.polygon(points)
        d.segment_label((slant, 0), (base + slant, 0), f"{base}")
        d.line((slant, 0), (slant, height), dashed=True)
        d.segment_label((slant, 0), (slant, height), f"{height}")
        d.right_angle((slant, 0), (base + slant, 0), (slant, height))
        
        return problem(
            question=f"Find the area of a parallelogram with base ${base}$ units and height ${height}$ units.",
            answer=area,
            difficulty=(1300, 1400),
            topic="geometry/area_of_polygons",
            solution=steps(
                f"Area of parallelogram = base × height",
                f"$A = {base} \\times {height}$",
                f"$A = {area}$ square units"
            ),
            image=d.render()
        )
    
    elif problem_type == 'trapezoid':
        base1 = randint(8, 18)
        base2 = randint(4, base1 - 2)
        height = randint(5, 12)
        area = Rational((base1 + base2) * height, 2)
        
        offset = (base1 - base2) / 2
        d = Diagram(width=300, height=220, padding=40)
        points = [(0, 0), (base1, 0), (base1 - offset, height), (offset, height)]
        d.polygon(points)
        d.segment_label((0, 0), (base1, 0), f"{base1}")
        d.segment_label((offset, height), (base1 - offset, height), f"{base2}")
        d.line((offset, 0), (offset, height), dashed=True)
        d.segment_label((offset, 0), (offset, height), f"{height}")
        d.right_angle((offset, 0), (base1, 0), (offset, height))
        
        return problem(
            question=f"Find the area of a trapezoid with bases ${base1}$ units and ${base2}$ units, and height ${height}$ units.",
            answer=area,
            difficulty=(1400, 1500),
            topic="geometry/area_of_polygons",
            solution=steps(
                f"Area of trapezoid = $\\frac{{1}}{{2}} \\times$ (base₁ + base₂) $\\times$ height",
                f"$A = \\frac{{1}}{{2}} \\times ({base1} + {base2}) \\times {height}$",
                f"$A = \\frac{{1}}{{2}} \\times {base1 + base2} \\times {height}$",
                f"$A = \\frac{{{(base1 + base2) * height}}}{{2}}$",
                f"$A = {latex(area)}$ square units"
            ),
            image=d.render()
        )
    
    else:  # composite shape
        rect_length = randint(8, 16)
        rect_width = randint(6, 12)
        tri_base = rect_length
        tri_height = randint(4, 10)
        
        rect_area = rect_length * rect_width
        tri_area = Rational(tri_base * tri_height, 2)
        total_area = rect_area + tri_area
        
        d = Diagram(width=300, height=250, padding=40)
        rect_points = [(0, 0), (rect_length, 0), (rect_length, rect_width), (0, rect_width)]
        d.polygon(rect_points)
        tri_points = [(0, rect_width), (rect_length, rect_width), (rect_length/2, rect_width + tri_height)]
        d.polygon(tri_points)
        d.segment_label((0, 0), (rect_length, 0), f"{rect_length}")
        d.segment_label((rect_length, 0), (rect_length, rect_width), f"{rect_width}")
        d.line((rect_length/2, rect_width), (rect_length/2, rect_width + tri_height), dashed=True)
        d.segment_label((rect_length/2, rect_width), (rect_length/2, rect_width + tri_height), f"{tri_height}")
        
        return problem(
            question=f"Find the total area of the composite shape: a rectangle with dimensions ${rect_length}$ by ${rect_width}$ units, topped with a triangle of base ${tri_base}$ units and height ${tri_height}$ units.",
            answer=total_area,
            difficulty=(1500, 1600),
            topic="geometry/area_of_polygons",
            solution=steps(
                f"Find the area of the rectangle: $A_{{rect}} = {rect_length} \\times {rect_width} = {rect_area}$ square units",
                f"Find the area of the triangle: $A_{{tri}} = \\frac{{1}}{{2}} \\times {tri_base} \\times {tri_height} = {latex(tri_area)}$ square units",
                f"Total area = $A_{{rect}} + A_{{tri}}$",
                f"$A = {rect_area} + {latex(tri_area)} = {latex(total_area)}$ square units"
            ),
            image=d.render()
        )

emit(generate())

File: src/geometry_area_of_polygons_medium.jl
# geometry - area_of_polygons (medium)
# Generated: 2026-03-08T20:07:58.285949

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script begin
    set_topic!("geometry/area_of_polygons")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Rectangle with diagonal - find area
        width = randint(5, 15)
        height = randint(5, 15)
        area = width * height
        diagonal = sqrt(width^2 + height^2)
        
        d = DiagramObj(600, 400)
        polygon!(d, [(50, 50), (50 + 30*width, 50), (50 + 30*width, 50 + 30*height), (50, 50 + 30*height)], 
                 fill="lightblue", stroke="black", stroke_width=2)
        line!(d, 50, 50, 50 + 30*width, 50 + 30*height, stroke="red", stroke_width=2, dash="5,5")
        segment_label!(d, 50, 50 - 20, 50 + 30*width, 50 - 20, "$(width)", position=:above)
        segment_label!(d, 50 + 30*width + 20, 50, 50 + 30*width + 20, 50 + 30*height, "$(height)", position=:right)
        text!(d, (50 + 30*width)/2, (50 + 30*height)/2 + 20, "d = $(round(diagonal, digits=2))", font_size=14)
        
        problem(
            question="A rectangle has a diagonal of length \$$(round(diagonal, digits=2))\$ units. If the width is \$$(width)\$ units and the height is \$$(height)\$ units, find the area of the rectangle.",
            answer=area,
            difficulty=(1200, 1400),
            solution=steps(
                "Given: width = $(width), height = $(height)",
                sol("Area formula", "A = width × height"),
                sol("Calculate", "A = $(width) × $(height) = $(area)")
            ),
            image=render(d),
            time=90
        )
        
    elseif problem_type == 2
        # Trapezoid area
        base1 = randint(8, 18)
        base2 = randint(5, base1 - 2)
        height = randint(6, 14)
        area = (base1 + base2) * height // 2
        
        d = DiagramObj(600, 400)
        offset = 30
        polygon!(d, [(50, 50 + 20*height), (50 + 20*base1, 50 + 20*height), 
                     (50 + 20*base1 - offset, 50), (50 + offset, 50)],
                 fill="lightgreen", stroke="black", stroke_width=2)
        segment_label!(d, 50, 50 + 20*height + 15, 50 + 20*base1, 50 + 20*height + 15, "$(base1)", position=:below)
        segment_label!(d, 50 + offset, 50 - 15, 50 + 20*base1 - offset, 50 - 15, "$(base2)", position=:above)
        line!(d, 50 + 20*base1 + 30, 50, 50 + 20*base1 + 30, 50 + 20*height, stroke="red", stroke_width=1, dash="3,3")
        segment_label!(d, 50 + 20*base1 + 35, 50, 50 + 20*base1 + 35, 50 + 20*height, "h = $(height)", position=:right)
        
        problem(
            question="Find the area of a trapezoid with parallel sides of length \$$(base1)\$ units and \$$(base2)\$ units, and a height of \$$(height)\$ units.",
            answer=area,
            difficulty=(1300, 1500),
            solution=steps(
                sol("Trapezoid area formula", "A = \\frac{1}{2}(b_1 + b_2)h"),
                sol("Substitute values", "A = \\frac{1}{2}($(base1) + $(base2))($(height))"),
                sol("Calculate", "A = \\frac{1}{2}($(base1 + base2))($(height)) = $(area)")
            ),
            image=render(d),
            time=120
        )
        
    elseif problem_type == 3
        # Parallelogram with side and height
        base = randint(10, 20)
        side = randint(8, 15)
        height = randint(6, min(side - 1, 12))
        area = base * height
        
        d = DiagramObj(600, 400)
        offset = 40
        polygon!(d, [(50, 50 + 20*height), (50 + 20*base, 50 + 20*height), 
                     (50 + 20*base + offset, 50), (50 + offset, 50)],
                 fill="lightyellow", stroke="black", stroke_width=2)
        line!(d, 50 + offset, 50, 50 + offset, 50 + 20*height, stroke="red", stroke_width=2, dash="4,4")
        segment_label!(d, 50, 50 + 20*height + 15, 50 + 20*base, 50 + 20*height + 15, "$(base)", position=:below)
        segment_label!(d, 50 + offset + 5, 50, 50 + offset + 5, 50 + 20*height, "h = $(height)", position=:right)
        segment_label!(d, 50 + offset - 15, 50 - 10, 50 - 15, 50 + 20*height - 10, "$(side)", position=:left)
        
        problem(
            question="A parallelogram has a base of \$$(base)\$ units, a side length of \$$(side)\$ units, and a perpendicular height of \$$(height)\$ units. Find the area of the parallelogram.",
            answer=area,
            difficulty=(1200, 1400),
            solution=steps(
                "The area of a parallelogram is base × height",
                sol("Formula", "A = b \\times h"),
                sol("Substitute", "A = $(base) \\times $(height) = $(area)")
            ),
            image=render(d),
            time=90
        )
        
    elseif problem_type == 4
        # Composite figure: rectangle with triangle on top
        rect_width = randint(10, 18)
        rect_height = randint(8, 14)
        tri_height = randint(5, 10)
        
        rect_area = rect_width * rect_height
        tri_area = rect_width * tri_height // 2
        total_area = rect_area + tri_area
        
        d = DiagramObj(600, 450)
        polygon!(d, [(50, 50 + 20*rect_height), (50 + 20*rect_width, 50 + 20*rect_height), 
                     (50 + 20*rect_width, 50), (50, 50)],
                 fill="lightblue", stroke="black", stroke_width=2)
        polygon!(d, [(50, 50), (50 + 20*rect_width, 50), (50 + 10*rect_width, 50 - 20*tri_height)],
                 fill="lightcoral", stroke="black", stroke_width=2)
        segment_label!(d, 50, 50 + 20*rect_height + 15, 50 + 20*rect_width, 50 + 20*rect_height + 15, "$(rect_width)", position=:below)
        segment_label!(d, 50 - 20, 50, 50 - 20, 50 + 20*rect_height, "$(rect_height)", position=:left)
        segment_label!(d, 50 + 20*rect_width + 20, 50 - 20*tri_height, 50 + 20*rect_width + 20, 50, "$(tri_height)", position=:right)
        
        problem(
            question="A figure consists of a rectangle with width \$$(rect_width)\$ units and height \$$(rect_height)\$ units, topped by a triangle with the same base and height \$$(tri_height)\$ units. Find the total area of the figure.",
            answer=total_area,
            difficulty=(1400, 1600),
            solution=steps(
                sol("Rectangle area", "A_1 = $(rect_width) \\times $(rect_height) = $(rect_area)"),
                sol("Triangle area", "A_2 = \\frac{1}{2} \\times $(rect_width) \\times $(tri_height) = $(tri_area)"),
                sol("Total area", "A = $(rect_area) + $(tri_area) = $(total_area)")
            ),
            image=render(d),
            time=150
        )
        
    else
        # Regular hexagon broken into triangles
        side = randint(6, 12)
        apothem_ratio = sqrt(3) / 2
        apothem = round(side * apothem_ratio, digits=2)
        perimeter = 6 * side
        area = round((perimeter * side * apothem_ratio) / 2, digits=2)
        
        d = DiagramObj(600, 600)
        cx, cy = 300, 300
        radius = 100
        angles = [60*i for i in 0:5]
        pts = [(cx + radius*cosd(a), cy + radius*sind(a)) for a in angles]
        polygon!(d, pts, fill="lavender", stroke="black", stroke_width=2)
        
        for i in 1:6
            line!(d, cx, cy, pts[i][1], pts[i][2], stroke="gray", stroke_width=1, dash="3,3")
        end
        
        point!(d, cx, cy, color="red", radius=4)
        segment_label!(d, cx, cy + 10, pts[1][1], pts[1][2] + 10, "$(side)", position=:above)
        text!(d, cx - 40, cy, "a = $(apothem)", font_size=12)
        
        problem(
            question="A regular hexagon has side length \$$(side)\$ units. The apothem (perpendicular distance from center to a side) is \$$(apothem)\$ units. Find the area of the hexagon.",
            answer=area,
            difficulty=(1500, 1700),
            solution=steps(
                sol("Perimeter", "P = 6 \\times $(side) = $(perimeter)"),
                sol("Area formula", "A = \\frac{1}{2} \\times P \\times a"),
                sol("Calculate", "A = \\frac{1}{2} \\times $(perimeter) \\times $(apothem) = $(area)")
            ),
            image=render(d),
            time=180
        )
    end
end